experimental/grpc_attestation/src/main.rs

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! gRPC Attestation Service implementation.
+//!
+//! gRPC Attestation Service performs remote attestation for each incoming streaming gRPC request.
+//!
+//! It is assumed that prior to the attestation workflow Service remotely attests its TEE platform
+//! key with the TEE firmware Provider. Provider sends back a signature rooted in some well-known
+//! root key (e.g. AMD root key), thus confirming the authenticity of the TEE platform key.
+//!
+//! Remote attestation is done as follows:
+//! - Client generates a Diffie-Hellman private/public key pair
+//! - Client sends a streaming gRPC request with a its public key to the Server
+//! - Server generates a Diffie-Hellman private/public key pair
+//! - Server uses its and Client's public keys to generate a session key
+//! - Server puts the hash of its public key into the TEE report request
+//!   - It means that when a TEE report is signed, the public key will also be signed
+//! - Service requests a TEE report from the TEE firmware
+//! - Server sends its public key, a TEE report and Provider's certificate to the Client
+//! - Client checks the validity of the TEE report
+//!   - It checks that the TEE report is signed by TEE Provider’s root key
+//!   - It checks that the public key hash from the TEE report is equal to the hash of the public
+//!     key presented in the server response
+//!   - It also extracts the TEE measurement from the TEE report and compares it to the expected one
+//!   - If the TEE report is not valid, then the Client closes the connection and aborts the
+//!     protocol
+//! - Client uses its and Server's public keys to generate a session key
```

Comment:
IIUC this handshake protocol does not bind the session key to the instance of the handshake protocol that creates it, opening up replay attacks.
For example: normally creating ephemeral keys for each session independently should mean that if one ephemeral key is compromised then that does not affect other sessions.
However here: compromising one server ephemeral key compromises all sessions in the future: simply replay the server response that attests the compromised ephemeral key.

The way handshake protocols generally deal with this: the server also signs a random challenge generated by the client. You can TAL at go/sc-key-provisioning#heading=h.wryz4qz7sacm for more details: tl;dr binding the ephemeral key to the transcript of the handshake so far which should include a challenge supplied by the peer you are establishing a session with.

Reply:
Thanks for pointing out!
Created an issue to implement a challenge-response stage of the protocol: https://github.com/project-oak/oak/issues/2105